docs/manual/inference.tex

```diff
@@ -126,22 +92,209 @@
 Cascade applies it.  Cascade works with any checker built on the Checker
 Framework.
 You can find installation instructions and a video tutorial at \url{https://github.com/reprogrammer/cascade}.
+% See last commit at https://github.com/reprogrammer/cascade/commits/master .
+Cascade was last updated in November 2014, so it might or might not work for you.
+
+\end{description}
+
+Except for one of the nullness inference tools, all these
+type inference tools are static analyses.  They analyze your program's
+source code, but they do not run your program.
 
 
 \sectionAndLabel{Whole-program inference}{whole-program-inference}
 
-Whole-program inference is an interprocedural inference that
+Whole-program inference
 infers types for fields, method parameters, and method return types that do not
 have a user-written qualifier (for the given type system).
-The inferred type qualifiers are inserted into
-your program.
+The inferred type qualifiers are output into annotation files.
 The inferred type is the most specific type that is compatible with all the
 uses in the program.  For example, the inferred type for a field is the
 least upper bound of the types of all the expressions that are assigned
 into the field.
 
+There are three scripts that you can use to run whole-program inference.
+Each has advantages and disadvantages, discussed below:
+
+\begin{itemize}
+    \item
+    To run whole-program inference on a single project without modifying its source code,
+    use the \<wpi.sh> script (Section~\ref{wpi-one}). This script can automatically understand
+    many Ant, Maven, and Gradle build files, so it requires little manual configuration.
+
+    \item
+    To run whole-program inference on many projects without modifying their source code
+    (say, when running it on projects from GitHub), use the \<wpi-many.sh> script (Section~\ref{wpi-many}).
+    This script can understand the same build files as \<wpi.sh>.
+
+    \item
+    If you want to insert the inferred annotations directly into a single
+    project's source code, use the \<infer-and-annotate.sh> script (Section~\ref{wpi-insert}).
+\end{itemize}
+
+These type inference scripts appear in the \<checker/bin/> directory.
+The remainder of this chapter describes them
+(Sections~\ref{wpi-one}--\ref{wpi-insert}), then concludes with discussion
+that applies to all of them.x
+
+
+\sectionAndLabel{Running whole-program inference on a single project}{wpi-one}
+
+A typical invocation of \<wpi.sh> is
+
+\begin{Verbatim}
+  wpi.sh -- --checker nullness
+\end{Verbatim}
+
+The result is a set of log files placed in the \<dljc-out> folder of the
+target project. The results of type-checking with each candidate set of
```

Comment:
Can you further explain "each candidate set of annotations"?  Is that from each iteration of type-checking?  Users only want to see the results of the final iteration.  Say how to see that.